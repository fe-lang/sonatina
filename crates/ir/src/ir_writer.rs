use std::io;

use smallvec::{Array, SmallVec};

use super::{BlockId, Function};
use crate::{
    module::{FuncRef, ModuleCtx},
    DataFlowGraph, InstId, Module, Value, ValueId,
};

pub struct ModuleWriter<'a> {
    pub module: &'a Module,
    pub dbg: &'a dyn DebugProvider,
}

impl<'a> ModuleWriter<'a> {
    pub fn new(module: &'a Module) -> Self {
        Self::with_debug_provider(module, &DEFAULT_PROVIDER)
    }

    pub fn with_debug_provider(module: &'a Module, dbg: &'a dyn DebugProvider) -> Self {
        Self { module, dbg }
    }

    pub fn write(&mut self, w: &mut impl io::Write) -> io::Result<()> {
        // Write target.
        writeln!(w, "target = {}", self.module.ctx.triple)?;
        writeln!(w)?;

        // Write struct types defined in the module.
        self.module.ctx.with_ty_store(|s| {
            let mut has_type_def = false;
            for s in s.all_struct_data() {
                has_type_def = true;
                s.write(w, &self.module.ctx)?;
                writeln!(w)?;
            }

            if has_type_def {
                writeln!(w)?;
            }

            io::Result::Ok(())
        })?;

        // Write module level global variables.
        self.module.ctx.with_gv_store(|s| {
            for gv in s.all_gv_data() {
                gv.write(w, &self.module.ctx)?;
                writeln!(w)?;
            }

            if !s.is_empty() {
                writeln!(w)?;
            }

            io::Result::Ok(())
        })?;

        for func_ref in self.module.func_store.funcs() {
            self.module.func_store.view(func_ref, |func| {
                let mut writer = FuncWriter::with_debug_provider(func, func_ref, self.dbg);
                writer.write(&mut *w)
            })?;
            writeln!(w)?;
        }

        Ok(())
    }

    pub fn dump_string(&mut self) -> String {
        let mut s = Vec::new();
        self.write(&mut s).unwrap();
        unsafe { String::from_utf8_unchecked(s) }
    }
}

pub struct FuncWriteCtx<'a> {
    pub func: &'a Function,
    pub func_ref: FuncRef,
    pub dbg: &'a dyn DebugProvider,
}

impl AsRef<ModuleCtx> for FuncWriteCtx<'_> {
    fn as_ref(&self) -> &ModuleCtx {
        self.module_ctx()
    }
}

impl<'a> FuncWriteCtx<'a> {
    pub fn with_debug_provider(
        func: &'a Function,
        func_ref: FuncRef,
        dbg: &'a dyn DebugProvider,
    ) -> Self {
        Self {
            func,
            func_ref,
            dbg,
        }
    }

    pub fn new(func: &'a Function, func_ref: FuncRef) -> Self {
        Self::with_debug_provider(func, func_ref, &DEFAULT_PROVIDER)
    }

    pub fn dfg(&self) -> &DataFlowGraph {
        &self.func.dfg
    }

    pub fn module_ctx(&self) -> &ModuleCtx {
        self.func.ctx()
    }
}

pub struct FuncWriter<'a> {
    ctx: FuncWriteCtx<'a>,
    level: u8,
}

impl<'a> FuncWriter<'a> {
    pub fn new(func_ref: FuncRef, func: &'a Function) -> Self {
        Self::with_debug_provider(func, func_ref, &DEFAULT_PROVIDER)
    }

    pub fn with_debug_provider(
        func: &'a Function,
        func_ref: FuncRef,
        dbg: &'a dyn DebugProvider,
    ) -> Self {
        let ctx = FuncWriteCtx::with_debug_provider(func, func_ref, dbg);
        Self { ctx, level: 0 }
    }

    pub fn write(&mut self, w: &mut impl io::Write) -> io::Result<()> {
        FunctionSignature.write(w, &self.ctx)?;
        writeln!(w, " {{")?;

        self.level += 1;

        for block in self.ctx.func.layout.iter_block() {
            self.write_block_with_inst(w, block)?;
            if !self.ctx.func.layout.is_last_block(block) {
                writeln!(w)?;
            }
        }

        self.level -= 1;
        writeln!(w, "}}")
    }

    pub fn dump_string(&mut self) -> String {
        let mut s = Vec::new();
        self.write(&mut s).unwrap();
        unsafe { String::from_utf8_unchecked(s) }
    }

    fn write_block_with_inst(&mut self, w: &mut impl io::Write, block: BlockId) -> io::Result<()> {
        self.indent(w)?;
        block.write(w, &self.ctx)?;

        self.enter(&mut *w)?;
        for inst in self.ctx.func.layout.iter_inst(block) {
            self.write_inst_in_block(w, inst)?;
        }
        self.leave();

        Ok(())
    }

    fn write_inst_in_block(&mut self, w: &mut impl io::Write, inst: InstId) -> io::Result<()> {
        self.indent(&mut *w)?;
        let inst_with_res = InstStatement(inst);
        inst_with_res.write(w, &self.ctx)?;
        writeln!(w)
    }

    fn indent(&self, w: &mut impl io::Write) -> io::Result<()> {
        write!(w, "{}", " ".repeat(self.level as usize * 4))
    }

    fn enter(&mut self, mut w: impl io::Write) -> io::Result<()> {
        self.level += 1;
        writeln!(w, ":")
    }

    fn leave(&mut self) {
        self.level -= 1;
    }
}

pub trait IrWrite<Ctx> {
    fn write<W>(&self, w: &mut W, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write;

    fn write_with_delim<W>(&self, w: &mut W, _delim: &str, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        self.write(w, ctx)
    }

    fn dump_string(&self, ctx: &Ctx) -> String {
        let mut bytes = Vec::new();
        self.write(&mut bytes, ctx).unwrap();
        unsafe { String::from_utf8_unchecked(bytes) }
    }

    fn has_content(&self) -> bool {
        true
    }
}

impl<Ctx, T, U> IrWrite<Ctx> for (T, U)
where
    T: IrWrite<Ctx>,
    U: IrWrite<Ctx>,
{
    fn write<W>(&self, w: &mut W, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        self.write_with_delim(w, " ", ctx)
    }

    fn write_with_delim<W>(&self, w: &mut W, delim: &str, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        write!(w, "(")?;
        self.0.write(w, ctx)?;
        if self.has_content() {
            write!(w, "{delim}")?;
            self.1.write(w, ctx)?;
        }
        write!(w, ")")
    }

    fn has_content(&self) -> bool {
        self.1.has_content() || self.has_content()
    }
}

impl<T, Ctx> IrWrite<Ctx> for [T]
where
    T: IrWrite<Ctx>,
{
    fn write<W>(&self, w: &mut W, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        self.write_with_delim(w, " ", ctx)
    }

    fn write_with_delim<W>(&self, w: &mut W, delim: &str, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        let mut iter = self.iter();
        let mut has_written = false;

        if let Some(value) = iter.next() {
            has_written |= value.has_content();
            value.write(w, ctx)?;
        }

        for value in iter {
            if has_written {
                write!(w, "{delim}")?;
            }
            value.write_with_delim(w, delim, ctx)?;
            has_written |= value.has_content();
        }

        Ok(())
    }

    fn has_content(&self) -> bool {
        self.iter().any(|x| x.has_content())
    }
}

impl<T, Ctx> IrWrite<Ctx> for Vec<T>
where
    T: IrWrite<Ctx>,
{
    fn write<W>(&self, w: &mut W, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        self.as_slice().write(w, ctx)
    }

    fn write_with_delim<W>(&self, w: &mut W, delim: &str, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        self.as_slice().write_with_delim(w, delim, ctx)
    }

    fn has_content(&self) -> bool {
        self.as_slice().has_content()
    }
}

impl<T, Ctx, const N: usize> IrWrite<Ctx> for SmallVec<[T; N]>
where
    [T; N]: Array<Item = T>,
    T: IrWrite<Ctx>,
{
    fn write<W>(&self, w: &mut W, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        self.as_slice().write(w, ctx)
    }

    fn write_with_delim<W>(&self, w: &mut W, delim: &str, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        self.as_slice().write_with_delim(w, delim, ctx)
    }

    fn has_content(&self) -> bool {
        self.as_slice().has_content()
    }
}

impl<T, Ctx> IrWrite<Ctx> for Option<T>
where
    T: IrWrite<Ctx>,
{
    fn write<W>(&self, w: &mut W, ctx: &Ctx) -> io::Result<()>
    where
        W: io::Write,
    {
        if let Some(content) = self {
            content.write(w, ctx)?;
        }

        Ok(())
    }

    fn has_content(&self) -> bool {
        self.as_ref().map(|x| x.has_content()).unwrap_or_default()
    }
}

#[derive(Clone, Copy)]
pub struct FunctionSignature;
impl IrWrite<FuncWriteCtx<'_>> for FunctionSignature {
    fn write<W>(&self, w: &mut W, ctx: &FuncWriteCtx) -> io::Result<()>
    where
        W: io::Write,
    {
        write!(w, "func ")?;
        ctx.func.sig.linkage().write(w, &ctx)?;
        write!(w, " %{}(", ctx.func.sig.name())?;
        let arg_values: SmallVec<[ValueWithTy; 8]> = ctx
            .func
            .arg_values
            .iter()
            .map(|value| ValueWithTy(*value))
            .collect();
        arg_values.write_with_delim(w, ", ", ctx)?;

        write!(w, ") -> ")?;
        ctx.func.sig.ret_ty().write(w, &ctx)
    }
}

#[derive(Clone, Copy)]
pub struct ValueWithTy(pub ValueId);
impl IrWrite<FuncWriteCtx<'_>> for ValueWithTy {
    fn write<W>(&self, w: &mut W, ctx: &FuncWriteCtx) -> io::Result<()>
    where
        W: io::Write,
    {
        let value = self.0;
        value.write(w, ctx)?;
        if let Value::Immediate { .. } = ctx.dfg().value(self.0) {
            Ok(())
        } else {
            let ty = ctx.dfg().value_ty(self.0);
            write!(w, ".")?;
            ty.write(w, ctx)
        }
    }
}

pub struct InstStatement(pub InstId);
impl IrWrite<FuncWriteCtx<'_>> for InstStatement {
    fn write<W>(&self, w: &mut W, ctx: &FuncWriteCtx) -> io::Result<()>
    where
        W: io::Write,
    {
        if let Some(result) = ctx.func.dfg.inst_result(self.0) {
            let result_with_ty = ValueWithTy(result);
            result_with_ty.write(w, ctx)?;
            write!(w, " = ")?;
        };

        self.0.write(w, ctx)?;
        write!(w, ";")
    }
}

pub trait DebugProvider {
    #[allow(unused)]
    fn value_name(&self, func: &Function, func_ref: FuncRef, value: ValueId) -> Option<&str> {
        None
    }
}

const DEFAULT_PROVIDER: DefaultDebugProvider = DefaultDebugProvider {};

struct DefaultDebugProvider {}
impl DebugProvider for DefaultDebugProvider {}
