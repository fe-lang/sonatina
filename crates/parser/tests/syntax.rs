use std::fmt::{self, Write};

use dir_test::{dir_test, Fixture};
use indenter::indented;
use ir::ir_writer::ModuleWriter;
use pest::{iterators::Pairs, Parser as _};
use sonatina_parser::{
    ast, parse_module,
    syntax::{Parser, Rule},
    Error,
};
mod common;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/syntax/func",
    glob: "*.sntn"
)]
fn test_func(fixture: Fixture<&str>) {
    test_rule(Rule::_functions, fixture)
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/syntax/module",
    glob: "*.sntn"
)]
fn test_module(fixture: Fixture<&str>) {
    test_rule(Rule::module, fixture)
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/syntax/module",
    glob: "*.sntn"
)]
fn test_module_ast(fixture: Fixture<&str>) {
    let module = ast::parse(fixture.content()).unwrap();
    snap_test!(format!("{:#?}", module), fixture.path(), Some("ast"));
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/syntax/module",
    glob: "*.sntn"
)]
fn test_module_ir(fixture: Fixture<&str>) {
    match parse_module(fixture.content()) {
        Ok(module) => {
            let mut w = ModuleWriter::with_debug_provider(&module.module, &module.debug);
            snap_test!(w.dump_string(), fixture.path(), Some("ir"));
        }
        Err(errs) => {
            for err in errs {
                let error_message = err.print_to_string(fixture.path(), fixture.content(), true);
                eprintln!("{}", error_message);
            }
            panic!("Failed to parse module");
        }
    }
}

fn test_rule(rule: Rule, fixture: Fixture<&str>) {
    match Parser::parse(rule, fixture.content()) {
        Ok(r) => {
            let s = format!("{:?}", PairsWrapper(r));
            snap_test!(s, fixture.path());
        }
        Err(err) => {
            report_error(err, &fixture);
            panic!();
        }
    }
}

fn report_error(err: pest::error::Error<Rule>, fixture: &Fixture<&str>) {
    let s = Error::SyntaxError(err).print_to_string(fixture.path(), fixture.content(), true);
    eprintln!("{s}");
}

struct PairsWrapper<'i>(Pairs<'i, Rule>);

impl fmt::Debug for PairsWrapper<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for pair in self.0.clone() {
            let rule = pair.as_rule();
            writeln!(f, "{rule:?} \"{}\"", pair.as_str(),)?;
            let pairs = pair.into_inner();
            if pairs.len() > 0 {
                write!(indented(f).with_str("  "), "{:?}", &PairsWrapper(pairs))?;
            }
        }
        Ok(())
    }
}
