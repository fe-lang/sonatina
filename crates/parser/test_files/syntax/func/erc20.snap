---
source: crates/parser/tests/syntax.rs
assertion_line: 53
input_file: test_files/syntax/func/erc20.sntn
---
function "func public %main() {
  block0:
    v0.i256 = evm_contract_size %constructor;
    v1.i256 = evm_contract_size %entrypoint;
    v2.i256 = add v0 v1;
    v3.i256 = get_function_ptr %entrypoint;
    evm_codecopy 0.i256 v2 v2;
}"
  function_signature "func public %main() "
    linkage "public"
    function_identifier "%main"
      function_name "main"
    function_params "()"
  block "block0:
      v0.i256 = evm_contract_size %constructor;
      v1.i256 = evm_contract_size %entrypoint;
      v2.i256 = add v0 v1;
      v3.i256 = get_function_ptr %entrypoint;
      evm_codecopy 0.i256 v2 v2;"
    block_ident "block0"
      block_number "0"
    stmt "v0.i256 = evm_contract_size %constructor;"
      assign_stmt "v0.i256 = evm_contract_size %constructor"
        value_declaration "v0.i256"
          value_name "v0"
          type_name "i256"
            primitive_type "i256"
        inst "evm_contract_size %constructor"
          inst_name "evm_contract_size"
            inst_identifier "evm_contract_size"
          inst_arg "%constructor"
            function_identifier "%constructor"
              function_name "constructor"
    stmt "v1.i256 = evm_contract_size %entrypoint;"
      assign_stmt "v1.i256 = evm_contract_size %entrypoint"
        value_declaration "v1.i256"
          value_name "v1"
          type_name "i256"
            primitive_type "i256"
        inst "evm_contract_size %entrypoint"
          inst_name "evm_contract_size"
            inst_identifier "evm_contract_size"
          inst_arg "%entrypoint"
            function_identifier "%entrypoint"
              function_name "entrypoint"
    stmt "v2.i256 = add v0 v1;"
      assign_stmt "v2.i256 = add v0 v1"
        value_declaration "v2.i256"
          value_name "v2"
          type_name "i256"
            primitive_type "i256"
        inst "add v0 v1"
          inst_name "add"
            inst_identifier "add"
          inst_arg "v0"
            value "v0"
              value_name "v0"
          inst_arg "v1"
            value "v1"
              value_name "v1"
    stmt "v3.i256 = get_function_ptr %entrypoint;"
      assign_stmt "v3.i256 = get_function_ptr %entrypoint"
        value_declaration "v3.i256"
          value_name "v3"
          type_name "i256"
            primitive_type "i256"
        inst "get_function_ptr %entrypoint"
          inst_name "get_function_ptr"
            inst_identifier "get_function_ptr"
          inst_arg "%entrypoint"
            function_identifier "%entrypoint"
              function_name "entrypoint"
    stmt "evm_codecopy 0.i256 v2 v2;"
      inst_stmt "evm_codecopy 0.i256 v2 v2"
        inst "evm_codecopy 0.i256 v2 v2"
          inst_name "evm_codecopy"
            inst_identifier "evm_codecopy"
          inst_arg "0.i256"
            value "0.i256"
              imm_number "0.i256"
                decimal "0"
                primitive_type "i256"
          inst_arg "v2"
            value "v2"
              value_name "v2"
          inst_arg "v2"
            value "v2"
              value_name "v2"
function "func private %constructor() {
  block0:
    v0.i128 = evm_caller;
    v2.i256 = evm_malloc 64.i256;
    v3.i256 = add v2 32.i256;
    evm_mstore v2 v0;
    evm_mstore v3 $MONEYBAG;
    v1.i256 = evm_keccak256 v2 64.i256;
    evm_sstore v1 1000000000000000000000.i256;
    v4.i256 = evm_malloc 32.i256;
    evm_mstore v4 0xDDF252AD1BE2C89B69C2B068FC378DAA952BA7F163C4A11628F55A4DF523B3EF.i256;
    v5.i128 = evm_caller;
    evm_log4 v4 32.i256 v0 v5 v1000000000000000000000;
}"
  function_signature "func private %constructor() "
    linkage "private"
    function_identifier "%constructor"
      function_name "constructor"
    function_params "()"
  block "block0:
      v0.i128 = evm_caller;
      v2.i256 = evm_malloc 64.i256;
      v3.i256 = add v2 32.i256;
      evm_mstore v2 v0;
      evm_mstore v3 $MONEYBAG;
      v1.i256 = evm_keccak256 v2 64.i256;
      evm_sstore v1 1000000000000000000000.i256;
      v4.i256 = evm_malloc 32.i256;
      evm_mstore v4 0xDDF252AD1BE2C89B69C2B068FC378DAA952BA7F163C4A11628F55A4DF523B3EF.i256;
      v5.i128 = evm_caller;
      evm_log4 v4 32.i256 v0 v5 v1000000000000000000000;"
    block_ident "block0"
      block_number "0"
    stmt "v0.i128 = evm_caller;"
      assign_stmt "v0.i128 = evm_caller"
        value_declaration "v0.i128"
          value_name "v0"
          type_name "i128"
            primitive_type "i128"
        inst "evm_caller"
          inst_name "evm_caller"
            inst_identifier "evm_caller"
    stmt "v2.i256 = evm_malloc 64.i256;"
      assign_stmt "v2.i256 = evm_malloc 64.i256"
        value_declaration "v2.i256"
          value_name "v2"
          type_name "i256"
            primitive_type "i256"
        inst "evm_malloc 64.i256"
          inst_name "evm_malloc"
            inst_identifier "evm_malloc"
          inst_arg "64.i256"
            value "64.i256"
              imm_number "64.i256"
                decimal "64"
                primitive_type "i256"
    stmt "v3.i256 = add v2 32.i256;"
      assign_stmt "v3.i256 = add v2 32.i256"
        value_declaration "v3.i256"
          value_name "v3"
          type_name "i256"
            primitive_type "i256"
        inst "add v2 32.i256"
          inst_name "add"
            inst_identifier "add"
          inst_arg "v2"
            value "v2"
              value_name "v2"
          inst_arg "32.i256"
            value "32.i256"
              imm_number "32.i256"
                decimal "32"
                primitive_type "i256"
    stmt "evm_mstore v2 v0;"
      inst_stmt "evm_mstore v2 v0"
        inst "evm_mstore v2 v0"
          inst_name "evm_mstore"
            inst_identifier "evm_mstore"
          inst_arg "v2"
            value "v2"
              value_name "v2"
          inst_arg "v0"
            value "v0"
              value_name "v0"
    stmt "evm_mstore v3 $MONEYBAG;"
      inst_stmt "evm_mstore v3 $MONEYBAG"
        inst "evm_mstore v3 $MONEYBAG"
          inst_name "evm_mstore"
            inst_identifier "evm_mstore"
          inst_arg "v3"
            value "v3"
              value_name "v3"
          inst_arg "$MONEYBAG"
            value "$MONEYBAG"
              global_value "$MONEYBAG"
                gv_identifier "$MONEYBAG"
                  gv_name "MONEYBAG"
    stmt "v1.i256 = evm_keccak256 v2 64.i256;"
      assign_stmt "v1.i256 = evm_keccak256 v2 64.i256"
        value_declaration "v1.i256"
          value_name "v1"
          type_name "i256"
            primitive_type "i256"
        inst "evm_keccak256 v2 64.i256"
          inst_name "evm_keccak256"
            inst_identifier "evm_keccak256"
          inst_arg "v2"
            value "v2"
              value_name "v2"
          inst_arg "64.i256"
            value "64.i256"
              imm_number "64.i256"
                decimal "64"
                primitive_type "i256"
    stmt "evm_sstore v1 1000000000000000000000.i256;"
      inst_stmt "evm_sstore v1 1000000000000000000000.i256"
        inst "evm_sstore v1 1000000000000000000000.i256"
          inst_name "evm_sstore"
            inst_identifier "evm_sstore"
          inst_arg "v1"
            value "v1"
              value_name "v1"
          inst_arg "1000000000000000000000.i256"
            value "1000000000000000000000.i256"
              imm_number "1000000000000000000000.i256"
                decimal "1000000000000000000000"
                primitive_type "i256"
    stmt "v4.i256 = evm_malloc 32.i256;"
      assign_stmt "v4.i256 = evm_malloc 32.i256"
        value_declaration "v4.i256"
          value_name "v4"
          type_name "i256"
            primitive_type "i256"
        inst "evm_malloc 32.i256"
          inst_name "evm_malloc"
            inst_identifier "evm_malloc"
          inst_arg "32.i256"
            value "32.i256"
              imm_number "32.i256"
                decimal "32"
                primitive_type "i256"
    stmt "evm_mstore v4 0xDDF252AD1BE2C89B69C2B068FC378DAA952BA7F163C4A11628F55A4DF523B3EF.i256;"
      inst_stmt "evm_mstore v4 0xDDF252AD1BE2C89B69C2B068FC378DAA952BA7F163C4A11628F55A4DF523B3EF.i256"
        inst "evm_mstore v4 0xDDF252AD1BE2C89B69C2B068FC378DAA952BA7F163C4A11628F55A4DF523B3EF.i256"
          inst_name "evm_mstore"
            inst_identifier "evm_mstore"
          inst_arg "v4"
            value "v4"
              value_name "v4"
          inst_arg "0xDDF252AD1BE2C89B69C2B068FC378DAA952BA7F163C4A11628F55A4DF523B3EF.i256"
            value "0xDDF252AD1BE2C89B69C2B068FC378DAA952BA7F163C4A11628F55A4DF523B3EF.i256"
              imm_number "0xDDF252AD1BE2C89B69C2B068FC378DAA952BA7F163C4A11628F55A4DF523B3EF.i256"
                hex "0xDDF252AD1BE2C89B69C2B068FC378DAA952BA7F163C4A11628F55A4DF523B3EF"
                primitive_type "i256"
    stmt "v5.i128 = evm_caller;"
      assign_stmt "v5.i128 = evm_caller"
        value_declaration "v5.i128"
          value_name "v5"
          type_name "i128"
            primitive_type "i128"
        inst "evm_caller"
          inst_name "evm_caller"
            inst_identifier "evm_caller"
    stmt "evm_log4 v4 32.i256 v0 v5 v1000000000000000000000;"
      inst_stmt "evm_log4 v4 32.i256 v0 v5 v1000000000000000000000"
        inst "evm_log4 v4 32.i256 v0 v5 v1000000000000000000000"
          inst_name "evm_log4"
            inst_identifier "evm_log4"
          inst_arg "v4"
            value "v4"
              value_name "v4"
          inst_arg "32.i256"
            value "32.i256"
              imm_number "32.i256"
                decimal "32"
                primitive_type "i256"
          inst_arg "v0"
            value "v0"
              value_name "v0"
          inst_arg "v5"
            value "v5"
              value_name "v5"
          inst_arg "v1000000000000000000000"
            value "v1000000000000000000000"
              value_name "v1000000000000000000000"
